htlcswitch/interfaces.go

```diff
@@ -138,8 +138,73 @@ type ChannelUpdateHandler interface {
 	// clean. This can be used with dynamic commitment negotiation or coop
 	// close negotiation which require a clean channel state.
 	ShutdownIfChannelClean() error
+
+	// EnableAdds sets the ChannelUpdateHandler state to allow
+	// UpdateAddHtlc's in the specified direction. It returns an error if
+	// the state already allowed those adds.
+	EnableAdds(direction LinkDirection) error
+
+	// DiableAdds sets the ChannelUpdateHandler state to allow
+	// UpdateAddHtlc's in the specified direction. It returns an error if
+	// the state already disallowed those adds.
+	DisableAdds(direction LinkDirection) error
+
+	// IsDraining returns true when UpdateAddHtlc's are disabled in the
+	// direction of the argument.
+	IsDraining(direction LinkDirection) bool
+
+	// OnFlushedOnce is a method that adds a hook to be called the next time
+	// the channel state reaches zero htlcs. This hook will only ever be
+	// called once. This function returns a FlushId that can be used to
+	// preemptively remove this hook.
+	OnFlushedOnce(func()) FlushHookID
```

Comment:
I think we are exposing too much implementation details of `link`, instead we should seek designs that can deepen the interface methods.

Reply:
I think we should have a conversation about what subsystems "own" each other. The way I've conceptualized everything right now is that we have:

Brontide <-> ChannelLink <-> Switch <-> ChannelLink <-> Brontide

and that each of these subsystems has a "peer" relationship with one another. That said, I can imagine and can endorse a design where the Switch "owns" the ChannelLinks. This makes sense to a degree because we already rely on the Switch to be able to look up the ChannelLink from the ChanID so there are already sketches of this relationship in the codebase today.

I don't think we should remove the API I put in here since I think it accurately describes what the role of the link is, but making the Brontide responsible for coordinating the lifecycle continuations is probably neither necessary nor ideal.

Does this match what you are thinking?

Reply:
Yep, I don't think we need to remove the API either, either on interface level or implementation level. But because `making the Brontide responsible for coordinating the lifecycle continuations is probably neither necessary nor ideal`, I'd say not expose `link` in `brontide`. Sharing some notes I've taken previously,
```
- When we init the shutdown:
	- request shutdown via `rpcServer.CloseChannel`
		- link is closed via `switch.CloseLink`, which creates a `htlcswitch.ChanClose` message that has a channel `Updates` that's used by the `chanCloser` to notify the channel is closed, then,
			- `switch` calls `s.cfg.LocalChannelClose`, mounted to `brontide.HandleLocalCloseChanReqs`
				- brontide calls `handleLocalCloseReq`:
					- 1. brontide creates a channel closer via `p.createChanCloser` and stores it in `p.activeChanCloses`, `htlcswitch.ChanClose` message is passed to this closer.
					- 2. brontide calls `chanCloser.ShutdownChan()`, in which,
						- `chanCloser` calls `initChanShutdown`, in which,
							- calls `c.cfg.DisableChannel`, that mounts to `ChanStatusMgr.RequestDisable`, in which,
								- `ChanStatusManager` calls `processDisableRequest`, in which,
									- it creates a channel update and calls `m.cfg.ApplyChannelUpdate`, mounts to `s.applyChannelUpdate` that sends a `ChannelUpdate` via the `authGossiper`
							- TODO calls it broadcasted with `nil` closing tx via `MarkCoopBroadcasted`, reason being to keep the `listchannels` rpc stay consistent <- need revisit
						- `chanCloser` sets its state to `closeShutdownInitiated`
					- 3. `brontide` calls `link.DisableAdds` which disables the outgoing link, which forbids us sending out htlcs to the peer.
					- 4. `brontide` sends the `shutdown` message to the peer
			- rpcserver now listens to `htlcswitch.ChanClose.Updates` which sends signals when the channel closing process has updates.
```

This is just the first part where we send out the `shutdown` message, notice how we jump from `switch` to `brontide` to `link`, not to mention the following handling on peer's `shutdown` message and `closing_signed` messages...